src/support.rs

```diff
@@ -21,47 +19,84 @@ pub trait FromRawSurface {
 }
 
 #[derive(Debug)]
-pub struct Writer<S: FromRawSurface + AsRef<Surface>, W: io::Write> {
-    pub surface: S,
-    writer: Box<W>,
+struct CallbackEnv<W> {
+    writer: W,
+    error: Option<io::Error>,
 }
 
-impl<S: FromRawSurface + AsRef<Surface>, W: io::Write> Writer<S, W> {
-    extern fn write_cb(writer: *mut c_void, data: *mut c_uchar, length: c_uint) -> cairo_status_t {
-        let mut writer: Box<W> = unsafe { Box::from_raw(writer as *mut _) };
-        let data = unsafe { slice::from_raw_parts(data, length as usize) };
+impl<W> CallbackEnv<W> {
+    fn new(writer: W) -> Box<Self> {
+        Box::new(CallbackEnv {
+            writer,
+            error: None,
+        })
+    }
+
+    fn as_void(&mut self) -> *mut c_void {
+        self as *mut Self as *mut c_void
+    }
+
+    unsafe fn from_void(ptr: &mut *mut c_void) -> &mut Self {
+        &mut *(*ptr as *mut Self)
+    }
 
-        let result = match writer.write_all(data) {
+    unsafe fn write(&mut self, data: *mut c_uchar, length: c_uint) -> cairo_status_t
+        where W: io::Write
+    {
+        let data = slice::from_raw_parts(data, length as usize);
+
+        let result = match self.writer.write_all(data) {
             Ok(_) => Status::Success,
-            Err(_) => Status::WriteError,
+            Err(e) => {
+                self.error = Some(e);
+                Status::WriteError
+            }
         };
 
-        mem::forget(writer);
         result.into()
     }
+}
+
+#[derive(Debug)]
+pub struct Writer<S: FromRawSurface + AsRef<Surface>, W: io::Write> {
+    pub surface: S,
+    callback_env: Box<CallbackEnv<W>>,
+}
+
+impl<S: FromRawSurface + AsRef<Surface>, W: io::Write> Writer<S, W> {
+    extern fn write_cb(mut env: *mut c_void, data: *mut c_uchar, length: c_uint) -> cairo_status_t {
+        unsafe {
+            // Safety: the type of `env` would matches `&mut *self.callback_env` in a `Writer` method.
```

Comment:
"would matches"?

Reply:
Fixed. (Then removed. This was more important when `EnvCallback<W>`

Reply:
Fixed. (Then removed. This was more important when `EnvCallback<W>` and `EnvCallback<&mut W>` were both in use.)